xds/internal/xdsclient/callback_serializer.go

```diff
@@ -0,0 +1,104 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package xdsclient
+
+import (
+	"context"
+	"sync"
+
+	"google.golang.org/grpc/internal/buffer"
+	"google.golang.org/grpc/internal/grpcsync"
+)
+
+// callbackSerializer provides a mechanism to schedule callbacks in a
+// synchronized manner. It provides a FIFO guarantee on the order of execution
+// of scheduled callbacks.
+//
+// New callbacks can be scheduled by invoking the Schedule() method. To cleanup
+// resources used by the serializer and to discard any queued callbacks, the
+// Close() method needs to be called.
+//
+// This type is safe for concurrent access.
+type callbackSerializer struct {
+	closedMu sync.Mutex
+	closed   bool
+
+	cancel    context.CancelFunc
+	done      *grpcsync.Event
+	callbacks *buffer.Unbounded
+}
+
+// newCallbackSerializer returns a new CallbackSerializer instance.
+func newCallbackSerializer() *callbackSerializer {
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
What if this accepted a context as a parameter, and we let the user cancel the context to `Close` it?  Is that better or worse?  I don't know precisely the context in which this will be used, but if there's already a context this could simplify things.  If not, it could make them a little more complex (maybe?).

Reply:
The callback serializer is currently used by `clientImpl` (to schedule watch callbacks for error conditions that happen even before it can determine the authority for the resource), and it is used by the `authority` type for scheduling watch callbacks for all cases (good response, bad response, resource is deleted etc).

Both these types don't have a context at this point in time. We _could_ add a context to these types, but that would probably be used only to be canceled at `Close()`, so that the callback serializer can cleanup. Creating a context, passing it to one type, and then canceling the context when we want to close that type seems cumbersome when compared to calling a `Close()` method.